### PR TITLE
feat: (Mobile) reset menu state on collapse 12477.

### LINKED ISSUE #12477: Menu does not reset itself after being closed on mobile view.
**Describe the bug**
When opening the menu in mobile view, whichever subsequent menus are opened remain when the menu is reopened.

**Tell us the version of Spartacus**
- Library version: 3.3

**To Reproduce**
1. Open Spartacus in mobile
2. Click on menu button
3. Go to any sub-category
4. Close the menu
5. Re-open the menu.

**Expected behavior**
Should start menu state from beginning with all submenus closed.

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser: Chrome

**Additional context**
I recommend making this fix togglable as a "feature" as some users may still prefer to maintain the menu state.

### PR DESCRIPTION
**Pull request by ren-tham**
Still being worked on.

### CODE REVIEW

File: projects/core/src/model/cms.model.ts [hunk 1]
```diff
@@ -151,6 +151,7 @@ export interface CmsNavigationComponent extends CmsComponent {
   notice?: string;
   showLanguageCurrency?: string;
   navigationNode?: CmsNavigationNode;
+  resetMenuOnCollapse?: boolean;
```
Comment (Zeyber) [inline projects/core/src/model/cms.model.ts]: Let's add some jsdoc comment here to explain what this flag does. Its something we all need to do more in general as its very helpful for a quick reference.

eg.
/**
 * Flag indicates whether to reset the state of menu navigation (ie. collapse all submenus) when the menu is closed.
 **/
resetMenuOnCollapse?: boolean;


Also, is it when the navigation panel is collapsed or closed?

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 2]
```diff
@@ -36,6 +41,11 @@ export class NavigationUIComponent implements OnDestroy {
    */
   iconType = ICON_TYPE;
 
+  /**
+   * Reset menu when collapsed
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Let's add more information here as to what "reset" means.

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 3]
```diff
@@ -36,6 +41,11 @@ export class NavigationUIComponent implements OnDestroy {
...
+   */
+  @Input() resetMenu: boolean;
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Let's also refer to this as `resetMenuOnCollapse` or `resetMenuOnClose`.

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 4]
```diff
@@ -54,10 +64,24 @@ export class NavigationUIComponent implements OnDestroy {
...
+    elemRef: ElementRef,
+    hamMenuService: HamburgerMenuService
+  );
+
   constructor(
     private router: Router,
     private renderer: Renderer2,
-    private elemRef: ElementRef
+    private elemRef: ElementRef,
+    @Optional() private hamMenuService?: HamburgerMenuService
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: For extensibility, can we make this service `protected` instead?

Comment (Zeyber) [reply projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Can we call this `hamburgerMenuService` instead of `hamMenuService`?

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 5]
```diff
@@ -71,6 +95,36 @@ export class NavigationUIComponent implements OnDestroy {
     );
   }
 
+  ngOnInit() {
+    if (this.resetMenu) this.resetOnMenuCollapse();
+  }
+
+  resetOnMenuCollapse() {
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Please add jsdocs to these methods to explain what they are doing. Please also add return value annotations (ie. `(): void`).

**Review by Zeyber — CHANGES_REQUESTED** [review]

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 6]
```diff
@@ -71,6 +95,42 @@ export class NavigationUIComponent implements OnDestroy {
...
+   * This method performs the action of resetting the menu (close all sub menus and return to main options)
+   * when the menu is closed.
+   */
+  resetOnMenuCollapse(): void {
+    this.subscriptions.add(
+      this.hamburgerMenuService.isExpanded
+        .pipe(distinctUntilChanged())
+        .subscribe((isExpanded: boolean) => {
+          //only react when expanded and proper number of nav node are present.
+          if (isExpanded && this.node.children?.length > 6) {
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Where does the number `6` come from? Better to set a constant and reference that so we can see what these numbers refer to (eg. this.node.children?.length > NODE_LIMIT, note: `NODE_LIMIT` is just an example var, I still don't know what it refers to).

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 7]
```diff
@@ -71,6 +95,42 @@ export class NavigationUIComponent implements OnDestroy {
     );
   }
 
+  /** During initialization of this component, we will check the resetMenuOnClose flag and attach a menu reset listener if needed.
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Please break this line to:
```
/**
 *  During....
 */
```

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 8]
```diff
@@ -71,6 +95,42 @@ export class NavigationUIComponent implements OnDestroy {
...
+          //only react when expanded and proper number of nav node are present.
+          if (isExpanded && this.node.children?.length > 6) {
+            if (this.openNodes?.length > 0) {
+              this.reinitalizeMenu();
+            }
+          }
+        })
+    );
+  }
+  reinitalizeMenu() {
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Please do the same docs and ts as for `resetOnMenuCollapse()`.

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 9]
```diff
@@ -71,6 +95,42 @@ export class NavigationUIComponent implements OnDestroy {
...
+    this.closeOpenedSubMenus();
+    this.renderer.removeClass(this.elemRef.nativeElement, 'is-open');
+  }
+  closeOpenedSubMenus() {
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Please do the same docs and ts as for `resetOnMenuCollapse()`.

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 10]
```diff
@@ -71,6 +95,42 @@ export class NavigationUIComponent implements OnDestroy {
...
+    let allClosed = false;
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Lets use typings wherever we can:
`let allClosed: boolean = false;`

**Review by Zeyber — CHANGES_REQUESTED** [review]

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 11]
```diff
@@ -71,6 +93,53 @@ export class NavigationUIComponent implements OnDestroy {
...
+    this.closeOpenedSubMenus();
+    this.renderer.removeClass(this.elemRef.nativeElement, 'is-open');
+  }
+
+  /**
+   * This method performs the action of closing all currently opened submenus and returns to the main navigation options menu.
+   */
+  closeOpenedSubMenus() {
+    let allClosed: boolean = false;
+    while (!allClosed) {
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Using a while loop can get code execution stuck if something unexpected happens. You should use something like a for loop that only gets called the number of times needed to avoid this risk.

Also, have you tried using the `clear()` function?

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 12]
```diff
@@ -36,6 +41,16 @@ export class NavigationUIComponent implements OnDestroy {
...
 
+  /**
+   * Flag indicates whether to reset the state of menu navigation (ie. Collapse all submenus) when the menu is closed.
+   */
+  @Input() resetMenuOnClose: boolean;
+
+  /**
+   * This value represents the number of navigation node options we expect.
+   */
+  @Input() numMenuNodes: number;
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Why is this required to be an input? Can't we reference the observable from within the class?

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 13]
```diff
@@ -36,6 +41,16 @@ export class NavigationUIComponent implements OnDestroy {
    */
   iconType = ICON_TYPE;
 
+  /**
+   * Flag indicates whether to reset the state of menu navigation (ie. Collapse all submenus) when the menu is closed.
+   */
+  @Input() resetMenuOnClose: boolean;
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts:40]: Instead of adding multiple inputs, lets have a configuration object as our input instead. Its tidier and easier to customize later. 

eg:

`navigation-ui.component.ts`
```
@Input config: NavigationUiConfig;
```

`navigation-ui-config.ts`
```
export interface NavigationUiConfig {
	  /**   
	  * Flag indicates whether to reset the state of menu navigation (ie. Collapse all submenus) when the menu is closed.   
	  */  
	  resetMenuOnClose?: boolean;
}
```

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 14]
```diff
@@ -71,6 +93,53 @@ export class NavigationUIComponent implements OnDestroy {
...
+   * This method performs the action of resetting the menu (close all sub menus and return to main options)
+   * when the menu is closed.
+   */
+  resetOnMenuCollapse(): void {
+    this.subscriptions.add(
+      this.hamburgerMenuService?.isExpanded
+        .pipe(distinctUntilChanged())
+        .subscribe((isExpanded: boolean) => {
+          //only react when expanded and proper number of nav node are present.
+          if (isExpanded && this.node.children?.length === this.numMenuNodes) {
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Do we really need to know how many nodes there are to close? Can't we just `clear()` open nodes and `updateClasses()`? The logic seems a bit complex but I might be missing something.

**Review by Zeyber — CHANGES_REQUESTED** [review]

File: projects/storefrontlib/src/cms-components/navigation/category-navigation/category-navigation.component.ts [hunk 15]
```diff
@@ -21,4 +23,18 @@ export class CategoryNavigationComponent {
     protected componentData: CmsComponentData<CmsNavigationComponent>,
     protected service: NavigationService
   ) {}
+
+  /**
+   * Returns navigation-ui config information.
+   */
+  getNavUIConfig(): Observable<NavigationUiConfig> {
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/category-navigation/category-navigation.component.ts]: Usually we pass configurations through the module via `provideConfig()` or `provideDefaultConfig()`. So in this case it would be `category-navigation.module.ts` or whichever imports the component.

File: projects/storefrontlib/src/cms-components/navigation/category-navigation/category-navigation.module.ts [hunk 16]
```diff
@@ -11,6 +11,9 @@ import { CategoryNavigationComponent } from './category-navigation.component';
       cmsComponents: {
         CategoryNavigationComponent: {
           component: CategoryNavigationComponent,
+          data: {
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/category-navigation/category-navigation.module.ts:14]: You can use another `provideDefaultConfig()` to import this configuration and remove it from component data.

`in separate default config file for component`
```
export const defaultCategoryNavigationConfig: NavigationUiConfig = {
	resetMenuOnClose: true
}
```

`in component module`
```
...
providers: [
	provideDefaultConfig(defaultCategoryNavigationConfig),
	...
```

File: projects/storefrontlib/src/cms-components/navigation/category-navigation/category-navigation.component.html [hunk 17]
```diff
@@ -1,6 +1,7 @@
 <cx-navigation-ui
   *ngIf="data$ | async as data"
   [node]="node$ | async"
+  [config]="getNavUIConfig() | async"
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/category-navigation/category-navigation.component.html]: We can remove this when we import the config by a service.

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.spec.ts [hunk 18]
```diff
@@ -24,6 +25,10 @@ class MockGenericLinkComponent {
   @Input() title: string;
 }
 
+const mockConfig: NavigationUiConfig = {
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.spec.ts]: Please add unit tests.

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 19]
```diff
@@ -71,6 +89,41 @@ export class NavigationUIComponent implements OnDestroy {
...
+  /**
+   * This method performs the action of resetting the menu (close all sub menus and return to main options)
+   * when the menu is closed.
+   */
+  resetOnMenuCollapse(): void {
+    this.subscriptions.add(
+      this.hamburgerMenuService?.isExpanded
+        .pipe(distinctUntilChanged())
+        .subscribe((isExpanded: boolean) => {
+          if (isExpanded) {
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Why is `isExpanded` not enough? It is possible to be expanded without open nodes? If so, can these conditions be in the same IF statement with `&&`?

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 20]
```diff
@@ -71,6 +89,41 @@ export class NavigationUIComponent implements OnDestroy {
...
+        })
+    );
+  }
+
+  /**
+   * This method performs the actions required to reset the state of the menu and reset any visual components.
+   */
+  reinitalizeMenu() {
+    this.clear();
+    this.renderer.removeClass(this.elemRef.nativeElement, 'is-open');
```
Comment (Zeyber) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: `this.clear()` doesn't work here?

Ren: excluding line 124, causes the menu to still remain in the menu that was last opened.

**Review by Zeyber — CHANGES_REQUESTED** [review]

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 21]
```diff
@@ -71,6 +89,41 @@ export class NavigationUIComponent implements OnDestroy {
...
+
+  /**
+   * This method performs the action of resetting the menu (close all sub menus and return to main options)
+   * when the menu is closed.
+   */
+  resetOnMenuCollapse(): void {
+    this.subscriptions.add(
+      this.hamburgerMenuService?.isExpanded
+        .pipe(distinctUntilChanged())
+        .subscribe((isExpanded: boolean) => {
```
Comment (Pucek9) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]:
```suggestion
        .pipe(
          distinctUntilChanged(),
          filter(Boolean),
        ).subscribe(
```
sorry for multiple comments for one fragment but there was some github bug which unallowed to suggest that part I wanted

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 22]
```diff
@@ -71,6 +89,41 @@ export class NavigationUIComponent implements OnDestroy {
...
+          if (isExpanded) {
+            this.reinitalizeMenu();
+          }
```
Comment (Pucek9) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]:
```suggestion          
          this.reinitalizeMenu();          
```

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 23]
```diff
@@ -71,6 +89,41 @@ export class NavigationUIComponent implements OnDestroy {
...
+  }
+
+  /**
+   * This method performs the action of resetting the menu (close all sub menus and return to main options)
+   * when the menu is closed.
+   */
+  resetOnMenuCollapse(): void {
+    this.subscriptions.add(
+      this.hamburgerMenuService?.isExpanded
+        .pipe(distinctUntilChanged())
```
Comment (Pucek9) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]:
```suggestion
```

File: projects/storefrontlib/src/cms-components/navigation/category-navigation/category-navigation.module.ts [hunk 24]
```diff
@@ -3,6 +3,7 @@ import { NgModule } from '@angular/core';
 import { CmsConfig, provideDefaultConfig } from '@spartacus/core';
 import { NavigationModule } from '../navigation/navigation.module';
 import { CategoryNavigationComponent } from './category-navigation.component';
+import { defaultCategoryNavigationConfig } from '../navigation/config/navigation.config';
```
Comment (Pucek9) [inline projects/storefrontlib/src/cms-components/navigation/category-navigation/category-navigation.module.ts]:
```suggestion
import { defaultNavigationConfig } from '../navigation/config/default-navigation.config';
```
or
```suggestion
import { defaultCategoryNavigationConfig } from '../navigation/config/default-category-navigation.config';
```
or
```suggestion
import { defaultCategoryNavigationUIConfig } from '../navigation/config/default-category-navigation-ui.config';
```

File: projects/storefrontlib/src/cms-components/navigation/index.ts [hunk 25]
```diff
@@ -9,6 +9,8 @@ export * from './navigation/navigation-ui.component';
 export * from './navigation/navigation.component';
 export * from './navigation/navigation.module';
 export * from './navigation/navigation.service';
+export * from './navigation/config/navigation-ui-config';
+export * from './navigation/config/navigation.config';
```
Comment (Pucek9) [inline projects/storefrontlib/src/cms-components/navigation/index.ts]:
```suggestion
export * from './navigation/config/default-navigation.config';
```
or
```suggestion
export * from './navigation/config/default-category-navigation.config';
```
or
```suggestion
export * from './navigation/config/default-category-navigation-ui.config';
```

File: projects/storefrontlib/src/cms-components/navigation/navigation/config/navigation.config.ts [hunk 26]
```diff
@@ -0,0 +1,5 @@
+import { NavigationUiConfig } from './navigation-ui-config';
+
+export const defaultCategoryNavigationConfig: NavigationUiConfig = {
```
Comment (Pucek9) [inline projects/storefrontlib/src/cms-components/navigation/navigation/config/navigation.config.ts]: I think our filename should be a reflection of class/interface/variable name, so please rename file to `default-category-navigation.config.ts`

or change

```suggestion
export const defaultNavigationConfig: NavigationUiConfig = {
```
or even
```suggestion
export const defaultCategorNavigationUIConfig: NavigationUiConfig = {
```
\+ rename file

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.spec.ts [hunk 27]
```diff
@@ -237,5 +255,21 @@ describe('Navigation UI Component', () => {
       );
       expect(child.length).toEqual(7);
     });
+
+    it('should reinitialize menu, when menu is expanded', () => {
+      let config: NavigationUiConfig = {
```
Comment (Pucek9) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.spec.ts]: please use `const` for variables which will not be changed, or just initialize 
```ts
navigationComponent.config = { resetMenuOnClose: true };
```

**Review by Pucek9 — DISMISSED** [review]

File: projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts [hunk 28]
```diff
@@ -54,10 +60,18 @@ export class NavigationUIComponent implements OnDestroy {
...
+   */
+  constructor(router: Router, renderer: Renderer2, elemRef: ElementRef);
+
   constructor(
     private router: Router,
     private renderer: Renderer2,
-    private elemRef: ElementRef
+    private elemRef: ElementRef,
+    @Optional() private config?: NavigationUiConfig,
+    @Optional() protected hamburgerMenuService?: HamburgerMenuService
```
Comment (marlass) [inline projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts]: Why not make it required params in this contructor?

File: projects/schematics/src/migrations/4_0/constructor-deprecations/data/navigation-ui.component.migration.ts [hunk 29]
```diff
@@ -0,0 +1,24 @@
...
+  NAVIGATION_UI_COMPONENT,
+  NAVIGATION_UI_CONFIG,
+} from '../../../../shared/constants';
+import { ConstructorDeprecation } from '../../../../shared/utils/file-utils';
+
+export const NAVIGATION_UI_COMPONENT_MIGRATION: ConstructorDeprecation = {
+  //projects/storefrontlib/src/cms-components/navigation/navigation/navigation-ui.component.ts
+  class: NAVIGATION_UI_COMPONENT,
+  importPath: SPARTACUS_STOREFRONTLIB,
+  deprecatedParams: [],
```
Comment (marlass) [inline projects/schematics/src/migrations/4_0/constructor-deprecations/data/navigation-ui.component.migration.ts]: To make this migration work the deprecatedParams must describe shape of the previous contructor (router, renderer2, elementRef)

**Review by marlass — CHANGES_REQUESTED** [review]

**Review by marlass — DISMISSED** [review]

**Review by marlass — APPROVED** [review]

**Review by Zeyber — DISMISSED** [review]

**Review by marlass — CHANGES_REQUESTED** [review]
After looking closely at this PR I don't think we should use Config for the steering behaviour of this component.

Can we have this configuration as a protected property in HamburgerMenuService?

If customers would like to change the default they would only need to extend this service and change the value of this property.

**Edit**: We are thinking that Cms Component config would be better for that

**Review by marlass — APPROVED** [review]
